### PR TITLE
add ONNX support for swin transformer

### PR DESCRIPTION
**Pull request by bibhabasumohapatra**
# What does this PR do?

Addresses #16308

## Before submitting

- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: docs/source/en/serialization.mdx [hunk 1]
```diff
@@ -91,6 +91,7 @@ Ready-made configurations include the following architectures:
 - XLM-RoBERTa
 - XLM-RoBERTa-XL
 - YOLOS
+- Swin
```
Comment (ChainYo) [inline docs/source/en/serialization.mdx:94]: Could you put it in alphabetical order, please?

File: src/transformers/models/swin/configuration_swin.py [hunk 2]
```diff
@@ -145,3 +151,20 @@ def __init__(
         # we set the hidden_size attribute in order to make Swin work with VisionEncoderDecoderModel
         # this indicates the channel dimension after the last stage of the model
         self.hidden_size = int(embed_dim * 2 ** (len(depths) - 1))
+
+
+class SwinOnnxConfig(OnnxConfig):
+
+    torch_onnx_minimum_version = version.parse("1.11")
```
Comment (ChainYo) [inline src/transformers/models/swin/configuration_swin.py:158]: Is there any reason you fixed a minimal version for this configuration?

**Comment by ChainYo** [comment]
Hey, @bibhabasumohapatra, Thanks for contributing to ONNX Config support. The PR looks almost good.

Could you run `make fix-copies` to fix the CI, as stated in the CI error statement?

File: src/transformers/models/swin/configuration_swin.py (hunk 2 above)
Comment (bibhabasumohapatra) [reply src/transformers/models/swin/configuration_swin.py:158]: I don't know the basic requirement, it's the same in every other Vision Onnx config-  without this package.version importing styling error in other Vit, BEiT, data2vec. So I maintained the same.

Comment (ChainYo) [reply src/transformers/models/swin/configuration_swin.py:158]: Ok, it makes sense!

**Comment by ChainYo** [comment]
To avoid your work from falling into limbo, I will ping @lewtun and @sgugger.

**Comment by sgugger** [comment]
@lewtun could have a look at this PR?

**Review by sgugger — APPROVED** [review]

File: src/transformers/models/swin/configuration_swin.py (hunk 2 above)
Comment (lewtun) [reply src/transformers/models/swin/configuration_swin.py:158]: Yes, we found that the PyTorch ONNX exporter had various issues for `torch <= 1.10.x` which is why it's fixed this way for all image models :)

File: src/transformers/models/swin/configuration_swin.py [hunk 3]
```diff
@@ -145,3 +151,20 @@ def __init__(
...
+
+    @property
+    def inputs(self) -> Mapping[str, Mapping[int, str]]:
+        return OrderedDict(
+            [
+                ("pixel_values", {0: "batch", 1: "sequence"}),
```
Comment (lewtun) [inline src/transformers/models/swin/configuration_swin.py:164]: We recently refactored the dynamic axis name for image models - please replace with the following:

```suggestion
                ("pixel_values", {0: "batch", 1: "num_channels"}),
```

File: src/transformers/onnx/features.py [hunk 4]
```diff
@@ -457,6 +457,9 @@ class FeaturesManager:
             "object-detection",
             onnx_config_cls="models.yolos.YolosOnnxConfig",
         ),
+        "swin": supported_features_mapping(
```
Comment (lewtun) [inline src/transformers/onnx/features.py:460]: Please order this entry alphabetically, i.e. just before `t5` (note to self: should automate this)

**Review by lewtun — COMMENTED** [review]
Thanks for adding this export @bibhabasumohapatra and welcome to the 🤗 Transformers community! 

Overall this looks good and there's just a few small changes that we need to include before merging. Can you also check that the slow tests pass with:

```
RUN_SLOW=1 pytest tests/onnx/test_onnx_v2.py -k "swin" -v
```

Edit: you'll also need to rebase / merge with `main` to address the merge conflicts.

File: src/transformers/onnx/features.py (hunk 4 above)
Comment (bibhabasumohapatra) [reply src/transformers/onnx/features.py]: @lewtun I did that but : https://app.circleci.com/pipelines/github/huggingface/transformers/46997/workflows/627f2d41-0b49-4f91-bf83-da5c5db701e8/jobs/555035/parallel-runs/0/steps/0-111

is showing a styling error in other files, should I run `make style` on other files? and there's a merge conflict tests/onnx/test_onnx_v2.py

**Comment by lewtun** [comment]
Hi @bibhabasumohapatra, sorry just getting back to this PR - would you mind rebasing on `main` to resolve the merge conflicts and then pushing again to check the CI is green?

After that, I think this will be good to go!

**Pull request closed without merging**